pyteal/ast/router_test.py

```diff
@@ -936,3 +938,32 @@ def handle(deposit: pt.abi.PaymentTransaction):  # noqa: F811
     ap2, cs2, c2 = r2.compile_program(version=pt.compiler.MAX_PROGRAM_VERSION)
 
     assert (ap1, cs1, c1) == (ap2, cs2, c2)
+
+
+def test_router_build_idempotence():
+    on_completion_actions = pt.BareCallActions(
+        opt_in=pt.OnCompleteAction.call_only(pt.Log(pt.Bytes("optin call"))),
+        clear_state=pt.OnCompleteAction.call_only(pt.Approve()),
+    )
+    router = pt.Router("questionable", on_completion_actions)
+
+    approval1, clear1, contract1 = router.build_program(version=6)
+    approval2, clear2, contract2 = router.build_program(version=6)
+
+    assert contract1.dictify() == contract2.dictify()
+    approval1, clear1, contract1 = router.compile_program(version=6)
+    approval2, clear2, contract2 = router.compile_program(version=6)
+
+    @pt.ABIReturnSubroutine
+    def add(a: pt.abi.Uint64, b: pt.abi.Uint64, *, output: pt.abi.Uint64) -> pt.Expr:
+        return output.set(a.get() + b.get())
+
+    meth = router.add_method_handler(add)
+    assert meth.method_signature() == "add(uint64,uint64)uint64"
+
+    approval1, clear1, contract1 = router.compile_program(version=6)
+    approval2, clear2, contract2 = router.compile_program(version=6)
+
+    assert contract1.dictify() == contract2.dictify()
+    assert clear1 == clear2
+    assert approval1 == approval2
```

Comment:
line 969 was failing for two reasons:
1. Unintended statefulness in the router that kept increasing compilation size. Every time `Router.build()` was called, it re-appended the routing conditions.
2. class variable `ScratchSlot.nextSlotId` keeps being incremented to during the life of program and doesn't reset after each compilation. This means that expressions that are recreated during compilation (such as the routing of bare-app calls) will keep getting assigned higher slot id's. The compiler may or not assign them to the same _finalized_ slot id's because of this artifact. By reseting `ScratchSlot.nextSlotId` hopefully this reduces the non-determinism. However, great care should be taken to assess whether the change is safe.